### TEXT SEARCH

In [70]:
import numpy as np

In [71]:
import requests 

docs_url = 'https://github.com/alexeygrigorev/llm-rag-workshop/raw/main/notebooks/documents.json'
docs_response = requests.get(docs_url)
documents_raw = docs_response.json()

documents = []

for course in documents_raw:
    course_name = course['course']

    for doc in course['documents']:
        doc['course'] = course_name
        documents.append(doc)

In [72]:
documents[2]


{'text': "Yes, even if you don't register, you're still eligible to submit the homeworks.\nBe aware, however, that there will be deadlines for turning in the final projects. So don't leave everything for the last minute.",
 'section': 'General course-related questions',
 'question': 'Course - Can I still join the course after the start date?',
 'course': 'data-engineering-zoomcamp'}

In [73]:
import pandas as pd

In [74]:
# convert to DataFrame & order columns
df = pd.DataFrame(documents, columns=['course', 'section', 'question', 'text'])
df.head()

course                           section  \
0  data-engineering-zoomcamp  General course-related questions   
1  data-engineering-zoomcamp  General course-related questions   
2  data-engineering-zoomcamp  General course-related questions   
3  data-engineering-zoomcamp  General course-related questions   
4  data-engineering-zoomcamp  General course-related questions   

                                            question  \
0               Course - When will the course start?   
1  Course - What are the prerequisites for this c...   
2  Course - Can I still join the course after the...   
3  Course - I have registered for the Data Engine...   
4   Course - What can I do before the course starts?   

                                                text  
0  The purpose of this document is to capture fre...  
1  GitHub - DataTalksClub data-engineering-zoomca...  
2  Yes, even if you don't register, you're still ...  
3  You don't need it. You're accepted. You can al...  
4  You can start by installing and setting up all...

In [75]:
df[df.course == 'data-engineering-zoomcamp'].head()

course                           section  \
0  data-engineering-zoomcamp  General course-related questions   
1  data-engineering-zoomcamp  General course-related questions   
2  data-engineering-zoomcamp  General course-related questions   
3  data-engineering-zoomcamp  General course-related questions   
4  data-engineering-zoomcamp  General course-related questions   

                                            question  \
0               Course - When will the course start?   
1  Course - What are the prerequisites for this c...   
2  Course - Can I still join the course after the...   
3  Course - I have registered for the Data Engine...   
4   Course - What can I do before the course starts?   

                                                text  
0  The purpose of this document is to capture fre...  
1  GitHub - DataTalksClub data-engineering-zoomca...  
2  Yes, even if you don't register, you're still ...  
3  You don't need it. You're accepted. You can al...  
4  You can start by installing and setting up all...

In [76]:
docs_example = [
    "January course details, register now",
    "Course prerequisites listed in January catalog",
    "Submit January course homework by end of month",
    "Register for January course, no prerequisites",
    "January course setup: Python and Google Cloud"
]

In [77]:
from sklearn.feature_extraction.text import CountVectorizer
# to turn text into vectors

In [78]:
# initialize CountVectorizer
cv = CountVectorizer()
#cv = CountVectorizer(min_df=5) # min_df=5 means ignore words that appear in less than 5 documents

In [79]:
# fit the vectorizer on the example documents
cv.fit(docs_example)

CountVectorizer()

In [80]:
cv.get_feature_names_out().shape

(22,)

In [81]:
names = cv.get_feature_names_out()
names

array(['and', 'by', 'catalog', 'cloud', 'course', 'details', 'end', 'for',
       'google', 'homework', 'in', 'january', 'listed', 'month', 'no',
       'now', 'of', 'prerequisites', 'python', 'register', 'setup',
       'submit'], dtype=object)

In [82]:
# transform into matrix
X = cv.transform(docs_example)

In [83]:
X


<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 32 stored elements and shape (5, 22)>

In [84]:
X.todense()

matrix([[0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0,
         0],
        [0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0,
         0],
        [0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0,
         1],
        [0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0,
         0],
        [1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1,
         0]])

In [85]:
X.toarray()

array([[0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0],
       [0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0],
       [0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1],
       [0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0],
       [1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0]])

In [86]:
df_docs = pd.DataFrame(X.toarray(), columns=cv.get_feature_names_out()).T
df_docs

0  1  2  3  4
and            0  0  0  0  1
by             0  0  1  0  0
catalog        0  1  0  0  0
cloud          0  0  0  0  1
course         1  1  1  1  1
details        1  0  0  0  0
end            0  0  1  0  0
for            0  0  0  1  0
google         0  0  0  0  1
homework       0  0  1  0  0
in             0  1  0  0  0
january        1  1  1  1  1
listed         0  1  0  0  0
month          0  0  1  0  0
no             0  0  0  1  0
now            1  0  0  0  0
of             0  0  1  0  0
prerequisites  0  1  0  1  0
python         0  0  0  0  1
register       1  0  0  1  0
setup          0  0  0  0  1
submit         0  0  1  0  0

This representation is called BAG OF WORDS.
- Dont care about order
- sparse matrics

In [87]:
cv = CountVectorizer(stop_words='english') # to ignore common words like 'the', 'and', etc.
X = cv.fit_transform(docs_example)

names = cv.get_feature_names_out()

df_docs = pd.DataFrame(X.toarray(), columns=names).T
df_docs

0  1  2  3  4
catalog        0  1  0  0  0
cloud          0  0  0  0  1
course         1  1  1  1  1
details        1  0  0  0  0
end            0  0  1  0  0
google         0  0  0  0  1
homework       0  0  1  0  0
january        1  1  1  1  1
listed         0  1  0  0  0
month          0  0  1  0  0
prerequisites  0  1  0  1  0
python         0  0  0  0  1
register       1  0  0  1  0
setup          0  0  0  0  1
submit         0  0  1  0  0

In [88]:
cv = CountVectorizer(stop_words='english', min_df=5) 
X = cv.fit_transform(df.text)

names = cv.get_feature_names_out()

df_docs = pd.DataFrame(X.toarray(), columns=names).T
df_docs

0    1    2    3    4    5    6    7    8    9    ...  938  939  \
01          0    0    0    0    0    0    0    0    0    0  ...    0    0   
02          0    0    0    0    0    0    0    0    0    0  ...    0    0   
03          0    0    0    0    0    0    0    0    0    0  ...    0    0   
04          0    0    0    0    0    0    0    0    0    0  ...    0    0   
05          0    0    0    0    0    0    0    0    0    0  ...    0    0   
...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
yes         0    0    1    0    0    0    1    1    1    0  ...    0    0   
yml         0    0    0    0    0    0    0    0    0    0  ...    0    0   
youtube     0    0    0    0    0    0    0    0    0    1  ...    0    0   
zip         0    0    0    0    0    0    0    0    0    0  ...    0    0   
zoomcamp    0    1    0    0    0    2    0    0    0    0  ...    0    0   

          940  941  942  943  944  945  946  947  
01          0    0    0    0    0    0    0    0  
02          0    0    0    0    0    0    0    0  
03          0    0    0    0    0    0    0    0  
04          0    0    0    0    0    0    0    0  
05          0    0    0    0    0    0    0    0  
...       ...  ...  ...  ...  ...  ...  ...  ...  
yes         0    0    0    0    0    0    0    0  
yml         0    0    0    1    0    0    0    0  
youtube     0    0    0    0    0    0    0    0  
zip         0    0    0    0    0    0    0    0  
zoomcamp    0    0    0    0    0    0    0    1  

[1333 rows x 948 columns]

we have 900 documents

##### Instead of using countVectorizer we can use TfidfVectorizer to give more importance to non that frequent words (example yaml vs yes) -> USED FOR SCORING

In [89]:
from sklearn.feature_extraction.text import TfidfVectorizer


In [90]:
cv = TfidfVectorizer(stop_words='english', min_df=5)
X = cv.fit_transform(df.text)

names = cv.get_feature_names_out()

df_docs = pd.DataFrame(X.toarray(), columns=cv.get_feature_names_out()).T
df_docs.round(2)

0     1     2    3    4     5     6    7     8     9    ...  938  \
01        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
02        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
03        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
04        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
05        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
...       ...   ...   ...  ...  ...   ...   ...  ...   ...   ...  ...  ...   
yes       0.0  0.00  0.28  0.0  0.0  0.00  0.21  0.2  0.15  0.00  ...  0.0   
yml       0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
youtube   0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.15  ...  0.0   
zip       0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
zoomcamp  0.0  0.43  0.00  0.0  0.0  0.19  0.00  0.0  0.00  0.00  ...  0.0   

          939  940  941  942   943  944  945  946   947  
01        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
02        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
03        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
04        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
05        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
...       ...  ...  ...  ...   ...  ...  ...  ...   ...  
yes       0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
yml       0.0  0.0  0.0  0.0  0.11  0.0  0.0  0.0  0.00  
youtube   0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
zip       0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
zoomcamp  0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.15  

[1333 rows x 948 columns]

In [91]:
query = "I just discovered the course, is it too late to join?"

In [92]:
q = cv.transform([query])
q.toarray()

array([[0., 0., 0., ..., 0., 0., 0.]], shape=(1, 1333))

In [93]:
query_dict = dict(zip(names, q.toarray()[0]))
query_dict

{'01': np.float64(0.0),
 '02': np.float64(0.0),
 '03': np.float64(0.0),
 '04': np.float64(0.0),
 '05': np.float64(0.0),
 '06': np.float64(0.0),
 '09': np.float64(0.0),
 '10': np.float64(0.0),
 '100': np.float64(0.0),
 '11': np.float64(0.0),
 '12': np.float64(0.0),
 '127': np.float64(0.0),
 '13': np.float64(0.0),
 '14': np.float64(0.0),
 '15': np.float64(0.0),
 '16': np.float64(0.0),
 '17': np.float64(0.0),
 '19': np.float64(0.0),
 '1st': np.float64(0.0),
 '20': np.float64(0.0),
 '2019': np.float64(0.0),
 '2020': np.float64(0.0),
 '2021': np.float64(0.0),
 '2022': np.float64(0.0),
 '2023': np.float64(0.0),
 '2024': np.float64(0.0),
 '21': np.float64(0.0),
 '22': np.float64(0.0),
 '24': np.float64(0.0),
 '25': np.float64(0.0),
 '2pacx': np.float64(0.0),
 '30': np.float64(0.0),
 '35': np.float64(0.0),
 '403': np.float64(0.0),
 '42': np.float64(0.0),
 '50': np.float64(0.0),
 '5000': np.float64(0.0),
 '5431': np.float64(0.0),
 '5432': np.float64(0.0),
 '60': np.float64(0.0),
 '600': np.floa

In [94]:
doc_dict = dict(zip(names, X.toarray()[1]))
doc_dict

{'01': np.float64(0.0),
 '02': np.float64(0.0),
 '03': np.float64(0.0),
 '04': np.float64(0.0),
 '05': np.float64(0.0),
 '06': np.float64(0.0),
 '09': np.float64(0.0),
 '10': np.float64(0.0),
 '100': np.float64(0.0),
 '11': np.float64(0.0),
 '12': np.float64(0.0),
 '127': np.float64(0.0),
 '13': np.float64(0.0),
 '14': np.float64(0.0),
 '15': np.float64(0.0),
 '16': np.float64(0.0),
 '17': np.float64(0.0),
 '19': np.float64(0.0),
 '1st': np.float64(0.0),
 '20': np.float64(0.0),
 '2019': np.float64(0.0),
 '2020': np.float64(0.0),
 '2021': np.float64(0.0),
 '2022': np.float64(0.0),
 '2023': np.float64(0.0),
 '2024': np.float64(0.0),
 '21': np.float64(0.0),
 '22': np.float64(0.0),
 '24': np.float64(0.0),
 '25': np.float64(0.0),
 '2pacx': np.float64(0.0),
 '30': np.float64(0.0),
 '35': np.float64(0.0),
 '403': np.float64(0.0),
 '42': np.float64(0.0),
 '50': np.float64(0.0),
 '5000': np.float64(0.0),
 '5431': np.float64(0.0),
 '5432': np.float64(0.0),
 '60': np.float64(0.0),
 '600': np.floa

#### Lets rank the documents by smilarity with query

In [95]:
df_qd = pd.DataFrame([query_dict, doc_dict], index=['query', 'doc']).T

In [96]:
(df_qd['query'] * df_qd['doc']).sum()

np.float64(0.0)

dot product for multiplications

In [97]:
X.dot(q.T).toarray()

array([[0.48049682],
       [0.        ],
       [0.        ],
       [0.2083882 ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.17557272],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.15870689],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.09680922],
       [0.        ],
       [0.        ],
       [0.07529201],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.29986763],
       [0.10520675],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.27447476],
       [0.12828407],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.05163407],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.03156309],
       [0.04914818],
       [0.07138962],
       [0.        ],
       [0.04329773],
       [0.        ],
       [0.        ],
       [0.        ],
       [0.   

to find similarities between matrix and query

In [98]:
from sklearn.metrics.pairwise import cosine_similarity

In [99]:
score = cosine_similarity(X, q).flatten()

In [100]:
np.argsort(score)[-5:]
            # gives indices of documents in ascending order

array([ 22, 448, 449, 440,   0])

the documents above are the top 5 similar (the 22 is the one with more matching words, and the 448 the second). Lets look at the answers:

In [101]:
print(df.loc[np.argsort(score)[-5:], ['text']].head(5))

                                                  text
22   It's up to you which platform and environment ...
448  Here’s how you join a in Slack: https://slack....
449  Yes, you can. You won’t be able to submit some...
440  The process is automated now, so you should re...
0    The purpose of this document is to capture fre...


#### Lets implement a search across all fields

In [102]:
df.columns

Index(['course', 'section', 'question', 'text'], dtype='object')

In [103]:
fields = ['section', 'question', 'text']
transformers = {}
matrices = {}

for f in fields:
    cv = TfidfVectorizer(stop_words='english', min_df=3)
    X = cv.fit_transform(df[f])

    transformers[f] = cv
    matrices[f] = X

In [105]:
matrices

{'section': <Compressed Sparse Row sparse matrix of dtype 'float64'
 	with 3094 stored elements and shape (948, 67)>,
 'question': <Compressed Sparse Row sparse matrix of dtype 'float64'
 	with 4333 stored elements and shape (948, 562)>,
 'text': <Compressed Sparse Row sparse matrix of dtype 'float64'
 	with 26463 stored elements and shape (948, 2118)>}

now we have matrices and vectoricers for all fields, lets look across all fields.

In [104]:
transformers['text'].get_feature_names_out()

array(['001', '01', '02', ..., 'zones', 'zoom', 'zoomcamp'],
      shape=(2118,), dtype=object)

In [36]:
matrices['text']

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 26463 stored elements and shape (948, 2118)>

In [37]:
query = "I just singned up. Is it too late to join the course?"

In [38]:
q = transformers['text'].transform([query])
score = cosine_similarity(matrices['text'], q).flatten()

In [39]:
mask = (df.course == 'data-engineering-zoomcamp').values
score = score * mask
score[:10]

array([0.3336047 , 0.        , 0.        , 0.1328874 , 0.        ,
       0.        , 0.        , 0.12722114, 0.        , 0.        ])

In [40]:
import numpy as np

In [41]:
idx = np.argsort(-score)[:10]
idx

array([  0,  15,  22,  27,  38, 287,   3,   7, 113,  11])

In [42]:
score[idx]

array([0.3336047 , 0.23530268, 0.22668   , 0.1894954 , 0.16484429,
       0.13921764, 0.1328874 , 0.12722114, 0.1207499 , 0.10830554])

In [43]:
df.iloc[idx].text

0      The purpose of this document is to capture fre...
15     No, late submissions are not allowed. But if t...
22     It's up to you which platform and environment ...
27     You can do most of the course without a cloud....
38     You will have two attempts for a project. If t...
287    This error could result if you are using some ...
3      You don't need it. You're accepted. You can al...
7      Yes, we will keep all the materials after the ...
113    In the join queries, if we mention the column ...
11     No, you can only get a certificate if you fini...
Name: text, dtype: object

In [44]:
fields

['section', 'question', 'text']

In [45]:
query = "I just signed up. Is it too late to join the course?"

More score to question that to answer

In [46]:
boost = {'question': 3.0} # boost the 'question' field 3.0 times more than others

score = np.zeros(len(df))

for f in fields:
    b = boost.get(f, 1.0)
    q = transformers[f].transform([query])
    s = cosine_similarity(matrices[f], q).flatten()
    score = score + b * s

Add filters to just select one course

In [47]:
filters = {
    'course': 'data-engineering-zoomcamp'
}

for field, value in filters.items():
    mask = (df[field] == value).values # this creates a boolean mask to filter documents just for this course
    score = score * mask

In [48]:
idx = np.argsort(-score)[:10]
results = df.iloc[idx]
results.to_dict(orient='records')

[{'course': 'data-engineering-zoomcamp',
  'section': 'General course-related questions',
  'question': 'Course - What are the prerequisites for this course?',
  'text': 'GitHub - DataTalksClub data-engineering-zoomcamp#prerequisites'},
 {'course': 'data-engineering-zoomcamp',
  'section': 'General course-related questions',
  'question': 'How can we contribute to the course?',
  'text': 'Star the repo! Share it with friends if you find it useful ❣️\nCreate a PR if you see you can improve the text or the structure of the repository.'},
 {'course': 'data-engineering-zoomcamp',
  'section': 'General course-related questions',
  'question': 'Course - What can I do before the course starts?',
  'text': 'You can start by installing and setting up all the dependencies and requirements:\nGoogle cloud account\nGoogle Cloud SDK\nPython 3 (installed with Anaconda)\nTerraform\nGit\nLook over the prerequisites and syllabus to see if you are comfortable with these subjects.'},
 {'course': 'data-eng

#### ALL TOGETHER 

In [49]:
class TextSearch:

    def __init__(self, text_fields):
        self.text_fields = text_fields
        self.matrices = {}
        self.vectorizers = {}

    def fit(self, records, vectorizer_params={}):
        self.df = pd.DataFrame(records)

        for f in self.text_fields:
            cv = TfidfVectorizer(**vectorizer_params)
            X = cv.fit_transform(self.df[f])
            self.matrices[f] = X
            self.vectorizers[f] = cv

    def search(self, query, n_results=10, boost={}, filters={}):
        score = np.zeros(len(self.df))

        for f in self.text_fields:
            b = boost.get(f, 1.0)
            q = self.vectorizers[f].transform([query])
            s = cosine_similarity(self.matrices[f], q).flatten()
            score = score + b * s

        for field, value in filters.items():
            mask = (self.df[field] == value).values
            score = score * mask

        idx = np.argsort(-score)[:n_results]
        results = self.df.iloc[idx]
        return results.to_dict(orient='records')

In [106]:
index = TextSearch(text_fields=['section', 'question', 'text'])
index.fit(documents)
index.search("I just signed up. Is it too late to join the course?",
             n_results=5,
             boost={'question': 3.0},
             filters={'course': 'data-engineering-zoomcamp'})

[{'text': "Yes, even if you don't register, you're still eligible to submit the homeworks.\nBe aware, however, that there will be deadlines for turning in the final projects. So don't leave everything for the last minute.",
  'section': 'General course-related questions',
  'question': 'Course - Can I still join the course after the start date?',
  'course': 'data-engineering-zoomcamp'},
 {'text': "The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the course Telegram channel with announcements.\nDon’t forget to register in DataTalks.Club's Slack and join the channel.",
  'section': 'General course-related questions',
  'question': 'Course - When will the course start?',
  'course': 'data-engineerin

### VECTOR/ELASTIC SEARCH

1. singular value decomposition: reduce the amount of values preserve (kind of compression, is not same as original one, compression dimensionality)

In [55]:
from sklearn.decomposition import TruncatedSVD

In [56]:
X = matrices['text']
cv = transformers['text']

In [57]:
svd = TruncatedSVD(n_components=16)
X_emb = svd.fit_transform(X)

In [58]:
X_emb[0]

array([ 0.08800166, -0.07490436, -0.09972477,  0.05149999,  0.05324144,
       -0.05627538,  0.01932436,  0.03270877, -0.18597129,  0.33844399,
        0.03839972,  0.10547948, -0.11314111,  0.07403075,  0.03812843,
       -0.00083166])

In [59]:
query = 'I just signed up. Is it too late to join the course?'

Q = cv.transform([query])
Q_emb = svd.transform(Q)

In [60]:
Q_emb[0]

array([ 0.04353715, -0.03060544, -0.04402867,  0.01220586,  0.02600693,
       -0.0503102 ,  0.01047362,  0.02088429, -0.11013902,  0.18220372,
        0.03273129,  0.07892531, -0.07483855,  0.03949898,  0.04002517,
       -0.00629577])

In [61]:
np.dot(X_emb[0], Q_emb[0])

np.float64(0.12089765261898303)

In [62]:
score = cosine_similarity(X_emb, Q_emb).flatten()

In [63]:
idx = np.argsort(-score)[:10]

In [64]:
list(df.loc[idx].text)

['If you have submitted two projects (and peer-reviewed at least 3 course-mates’ projects for each submission), you will get the certificate for the course. According to the course coordinator, Alexey Grigorev, only two projects are needed to get the course certificate.\n(optional) David Odimegwu',
 'Yes, you can. You won’t be able to submit some of the homeworks, but you can still take part in the course.\nIn order to get a certificate, you need to submit 2 out of 3 course projects and review 3 peers’ Projects by the deadline. It means that if you join the course at the end of November and manage to work on two projects, you will still be eligible for a certificate.',
 'No, late submissions are not allowed. But if the form is still not closed and it’s after the due date, you can still submit the homework. confirm your submission by the date-timestamp on the Course page.y\nOlder news:[source1] [source2]',
 'The course videos are pre-recorded, you can start watching the course right now

In [65]:
from sklearn.decomposition import NMF

In [66]:
nmf = NMF(n_components=16)
X_emb = nmf.fit_transform(X)
X_emb[0]

array([0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.31122894,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        ])

In [67]:
Q = cv.transform([query])
Q_emb = nmf.transform(Q)
Q_emb[0]

array([0.        , 0.00114793, 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.17392236,
       0.        , 0.        , 0.        , 0.        , 0.0006448 ,
       0.        ])

In [68]:
score = cosine_similarity(X_emb, Q_emb).flatten()
idx = np.argsort(-score)[:10]
list(df.loc[idx].text)

["The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the course Telegram channel with announcements.\nDon’t forget to register in DataTalks.Club's Slack and join the channel.",
 'Please choose the closest one to your answer. Also do not post your answer in the course slack channel.',
 "Yes, even if you don't register, you're still eligible to submit the homeworks.\nBe aware, however, that there will be deadlines for turning in the final projects. So don't leave everything for the last minute.",
 'No, it’s not possible. The form is closed after the due date. But don’t worry, homework is not mandatory for finishing the course.',
 'Yes, you can. You won’t be able to submit some of the homeworks, but yo

In [69]:
import torch
from transformers import BertModel, BertTokenizer

tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
model = BertModel.from_pretrained("bert-base-uncased")
model.eval()  # Set the model to evaluation mode if not training

ModuleNotFoundError: No module named 'torch'

In [ ]:
texts = [
    "Yes, we will keep all the materials after the course finishes.",
    "You can follow the course at your own pace after it finishes"
]
encoded_input = tokenizer(texts, padding=True, truncation=True, return_tensors='pt')


In [ ]:
encoded_input

{'input_ids': tensor([[  101,  2748,  1010,  2057,  2097,  2562,  2035,  1996,  4475,  2044,
          1996,  2607, 12321,  1012,   102],
        [  101,  2017,  2064,  3582,  1996,  2607,  2012,  2115,  2219,  6393,
          2044,  2009, 12321,   102,     0]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0]])}

In [ ]:
with torch.no_grad():  # Disable gradient calculation for inference
    outputs = model(**encoded_input)
    hidden_states = outputs.last_hidden_state

In [ ]:
hidden_states.shape

torch.Size([2, 15, 768])

In [ ]:
sentence_embeddings = hidden_states.mean(dim=1)
sentence_embeddings.shape

torch.Size([2, 768])

In [ ]:
sentence_embeddings.numpy()

# note that if use a GPU, first you need to move your tensors to CPU
# sentence_embeddings_cpu = sentence_embeddings.cpu()

array([[ 0.35999233, -0.16072303,  0.35452363, ...,  0.04289245,
         0.03482292, -0.03822247],
       [ 0.17849916, -0.500025  ,  0.25277564, ..., -0.11413109,
        -0.3360847 ,  0.4109514 ]], dtype=float32)

In [ ]:
def make_batches(seq, n):
    result = []
    for i in range(0, len(seq), n):
        batch = seq[i:i+n]
        result.append(batch)
    return result

In [ ]:
from tqdm.auto import tqdm

In [ ]:
def compute_embeddings(texts, batch_size=8):
    text_batches = make_batches(texts, 8)
    
    all_embeddings = []
    
    for batch in tqdm(text_batches):
        encoded_input = tokenizer(batch, padding=True, truncation=True, return_tensors='pt')
    
        with torch.no_grad():
            outputs = model(**encoded_input)
            hidden_states = outputs.last_hidden_state
            
            batch_embeddings = hidden_states.mean(dim=1)
            batch_embeddings_np = batch_embeddings.cpu().numpy()
            all_embeddings.append(batch_embeddings_np)
    
    final_embeddings = np.vstack(all_embeddings)
    return final_embeddings

In [ ]:
embeddings = {}

In [ ]:
# fields = ['section', 'question', 'text']

for f in fields:
    print(f'computing embeddings for {f}...')
    embeddings[f] = compute_embeddings(df[f].tolist())

computing embeddings for section...


  0%|          | 0/119 [00:00<?, ?it/s]

computing embeddings for question...


  0%|          | 0/119 [00:00<?, ?it/s]

In [ ]:
import pickle

In [ ]:
with open('embeddings.bin', 'wb') as f_out:
    pickle.dump(embeddings)